main.go

```diff
@@ -187,11 +190,58 @@ func obfuscatedTypesPackage(path string) *types.Package {
 
 var cachedTypesPackages = make(map[string]*types.Package)
 
+// uniqueLineWriter sits underneath log.SetOutput to deduplicate log lines.
+// We log bits of useful information for debugging,
+// and logging the same detail twice is not going to help the user.
+// Duplicates are relatively normal, given names tend to repeat.
+type uniqueLineWriter struct {
+	out  io.Writer
+	seen map[string]bool
+}
+
+func (w *uniqueLineWriter) Write(p []byte) (n int, err error) {
+	if !flagDebug {
+		panic("unexpected use of uniqueLineWriter with -debug unset")
+	}
+	if bytes.Count(p, []byte("\n")) != 1 {
+		panic(fmt.Sprintf("log write wasn't just one line: %q", p))
+	}
+	if w.seen[string(p)] {
+		return len(p), nil
+	}
+	if w.seen == nil {
+		w.seen = make(map[string]bool)
+	}
+	w.seen[string(p)] = true
+	return w.out.Write(p)
+}
+
+// debugf is like log.Printf, but it is a no-op by default.
+// TODO(mvdan): remove once we use 1.18: https://github.com/golang/go/issues/47164
+func debugf(format string, args ...interface{}) {
+	if !flagDebug {
+		return
+	}
+	log.Printf(format, args...)
+}
+
+// debugSince is like time.Since but resulting in shorter output.
+// A build process takes at least hundreds of milliseconds,
+// so extra decimal points in the order of microseconds aren't meaningful.
+func debugSince(start time.Time) time.Duration {
+	return time.Since(start).Truncate(10 * time.Microsecond)
+}
+
 func main1() int {
 	if err := flagSet.Parse(os.Args[1:]); err != nil {
 		return 2
 	}
-	log.SetPrefix("[garble] ")
```

Comment:
I'm a bit torn on this, because often the prefix is repetitive. But I'll agree that it's best to be explicit, and the toolchain does print its own output too.